### DISCUSSION TITLE
Please allow to globally disable prefetching

### DISCUSSION THREAD

**Discussion by tacone** [vercel/next.js#24009] [Ideas]
### Describe the feature you'd like to request

Please allow `<Link>` prefetching to be globally disabled, or enabled one a one-to-one basis.

In development mode prefetching raises the noise to signal ratio, and makes it difficult to get a sense of the page footprint.

In production mode prefetching is not always desirable and Next should not try to outsmart the developer. 

Globally disabling prefetching currently involves [implementing your own Link element](https://web.dev/route-prefetching-in-nextjs/) which is cumbersome, or making sure that every <Link> in your app has prefetch set to false.

### Describe the solution you'd like

One solution may be to allow prefetching to be globally disabled by setting an entry in `next.config.js`. Setting such entry should set the default `prefetching` strategy of `<Links>` when a value is not set.

Possible values may be:
```
prefetch: 'auto'  // current behaviour
prefetch: 'off'  // prefetching defaults to false, but can be overriden by <Link prefetching
prefetch: 'force-off'  // disable prefetching, no matter what
```

An alternative solution may be to let the developer writing a JS function to determine if the fetch should happen or not.

### Describe alternatives you've considered

I found none.

**Comment by timneutkens** [comment]
> In development mode prefetching raises the noise to signal ratio, and makes it difficult to get a sense of the page footprint.

In development prefetching is disabled.

> Globally disabling prefetching currently involves implementing your own Link element which is cumbersome, or making sure that every in your app has prefetch set to false.

This comes down to just this code:

```
import Link from 'next/link'

export default function MyLink(props) {
  // defaults prefetch to false if `prefetch` is not true
  return <Link {...props} prefetch={props.prefetch ?? false}>
}
```

**Reply by hatsumatsu** [reply]
@timneutkens setting `prefetch={false}` still prefetches on mouseover without any way to disable it completely, right?
There seem to be use cases where this isn't ideal like #20521 or #23641 .

**Reply by bennettdams** [reply]
Now that `next/link` will be updated for `v13`, maybe this is a good time to add this feature as well (as it may be a breaking change)? Some people including me can't use `next/link` now at all, as even prefetching on hover can be very critical - e.g. in a navbar of data-heavy pages. A user just has to mindlessly strafe the mouse from left to right over the navbar and the navigation is blocked until all data is loaded...

**Reply by elie222** [reply]
Is there a nice TS version of this component?

**Reply by jakevollkommer** [reply]
TypeScript option:

```
import Link from 'next/link'
import { ComponentProps } from 'react'

export default function NoPrefetchLink(props: ComponentProps<typeof Link>) {
  return <Link {...props} prefetch={props.prefetch ?? false}>
    {props.children}
  </Link>
};
```

**Reply by abhishekmardiya** [reply]
```ts
import Link from "next/link";
import type { HTMLAttributeAnchorTarget, ReactNode } from "react";

interface NextLinkProps {
  href: string;
  target?: HTMLAttributeAnchorTarget;
  prefetch?: boolean;
  className?: string;
  children: ReactNode;
}

export const NextLink = ({
  href,
  target,
  prefetch,
  className,
  children,
}: NextLinkProps) => {
  return (
    <Link
      href={href || "/"}
      target={target || "_self"}
      prefetch={prefetch || false}
      className={className}
      scroll={true}
    >
      {children}
    </Link>
  );
};
```

**Comment by JamieMason** [comment]
What I've done is create a Babel plugin local to my project which adds `prefetch={false}` everywhere we use `<Link />` from 'next/link'.

1 - Create a file at eg `./babel/disable-nextjs-link-prefetching.js` or wherever you like:

```js
/**
 * Based on the docs at https://nextjs.org/docs/api-reference/next/link, the
 * only way to disable prefetching is to make sure every <Link /> has <Link
 * prefetch={false} />
 *
 * We don't want to create a wrapper Component or go around changing every
 * single <Link />, so we use this Babel Plugin to add them in at build-time.
 */
module.exports = function (babel) {
  const { types: t } = babel
  return {
    name: 'disable-link-prefetching',
    visitor: {
      JSXOpeningElement(path) {
        if (path.node.name.name === 'Link') {
          path.node.attributes.push(
            t.jSXAttribute(
              t.jSXIdentifier('prefetch'),
              t.JSXExpressionContainer(t.booleanLiteral(false)),
            ),
          )
        }
      },
    },
  }
}
```

2 - Add/modify `./.babelrc` to load your local plugin:

```json
{
  "presets": ["next/babel"],
  "plugins": ["./babel/disable-nextjs-link-prefetching"]
}
```

**Reply by urbanrotnik** [reply]
> What I've done is create a Babel plugin local to my project which adds `prefetch={false}` everywhere we use `<Link />` from 'next/link'.
> 
> 1 - Create a file at eg `./babel/disable-nextjs-link-prefetching.js` or wherever you like:
> 
> ```js
> /**
>  * Based on the docs at https://nextjs.org/docs/api-reference/next/link, the
>  * only way to disable prefetching is to make sure every <Link /> has <Link
>  * prefetch={false} />
>  *
>  * We don't want to create a wrapper Component or go around changing every
>  * single <Link />, so we use this Babel Plugin to add them in at build-time.
>  */
> module.exports = function (babel) {
>   const { types: t } = babel
>   return {
>     name: 'disable-link-prefetching',
>     visitor: {
>       JSXOpeningElement(path) {
>         if (path.node.name.name === 'Link') {
>           path.node.attributes.push(
>             t.jSXAttribute(
>               t.jSXIdentifier('prefetch'),
>               t.JSXExpressionContainer(t.booleanLiteral(false)),
>             ),
>           )
>         }
>       },
>     },
>   }
> }
> ```
> 
> 2 - Add/modify `./.babelrc` to load your local plugin:
> 
> ```json
> {
>   "presets": ["next/babel"],
>   "plugins": ["./babel/disable-nextjs-link-prefetching"]
> }
> ```

Does it prevent prefetching on hover also?

**Reply by JamieMason** [reply]
I'm not sure off the top of my head but if you check the docs for what `prefetch={false}` does that should clear it up. All the plug-in does is make sure that prop is always set.

EDIT: I was on mobile when I replied earlier, I see from your original comment that there's also a `prefetch: 'force-off'` – the babel plugin could be modified to set that value instead:

```diff
t.jSXAttribute(
  t.jSXIdentifier('prefetch'),
- t.JSXExpressionContainer(t.booleanLiteral(false)),
+ t.stringLiteral("force-off"),
),
```

## Examples on AST Explorer

- [Babel plugin to set `prefetch={false}` on next/link](https://astexplorer.net/#/gist/3d2cf0c0a3740f534546bf562c22760b/5790894a0354debaa7fbd5d8aeeffe7e33cb66f8)
- [Babel plugin to set `prefetch="force-off"` on next/link](https://astexplorer.net/#/gist/aa2dded95e6dad4df33b9592fc4f63e6/8fde931c9d19c1d73f2dea99cf55846df5f806fe)

**Reply by 0oooooooo0** [reply]
The above code changes all prefetch to false, but I made it to false only if there is no prefetch.

```
module.exports = function ({ types: t }) {
  return {
    name: 'nextjs-link-prefetch-default',
    visitor: {
      JSXOpeningElement({ node }) {
        if (node.name.name === 'Link') {
          const defaultValue = false
          const prefetchAttributes = node.attributes.filter(attribute => attribute.type === 'JSXAttribute' && attribute.name.name === 'prefetch')
          if (prefetchAttributes.length <= 0) {
            node.attributes.push(t.jSXAttribute(t.jSXIdentifier('prefetch'), t.JSXExpressionContainer(t.booleanLiteral(defaultValue))))
          }
        }
      },
    },
  }
}
```

**Reply by JamieMason** [reply]
Great @0oooooooo0, much better 👍

**Reply by 0oooooooo0** [reply]
By the way, this makes swc unusable. I'm thinking of another way 🥲

**Reply by JamieMason** [reply]
That would be great. I only know of this Babel way and you're right, using Babel locks you out of using SWC.

**Reply by Programming-Seungwan** [reply]
@JamieMason @0oooooooo0 Guys, Did you find ouy any way to apply prefetch option to Link components in overall nextJS project?

**Comment by bennettdams** [comment]


**Reply by NawarA** [reply]
No, you're wrong. That's a different request than this request.

This request has to do with having a global config, so you can set a flag as the default behavior, so devs can override default settings.

The one you say is a dupe, is talking about how `prefetch` has a specific behavior of "pre-fetching on mouse hover`. It says there's no way to disable that mouse hover pre-fetch, even when pre-fetch equals false,

These are two different discussions, and neither should be sidelined because it looks like its a "dupe". They're not dupes, two different requests 

❤️

**Reply by bennettdams** [reply]
Right, deleted my comment.

**Comment by BertrandBordage** [comment]
A workaround that I use for **disabling hover prefetch as well** is essentially this:
```typescript
import React from "react";
import { Router, withRouter } from "next/router";

type Props = {
  href: string;
  children: React.ReactNode;
  router: Router;
};

class LinkNoPrefetch extends React.PureComponent<Props> {
  onClick = async (event: React.MouseEvent) => {
    const { router, href } = this.props;
    event.preventDefault();
    await router.push(href);
  }

  render() {
    const { href, children } = this.props;
    return (
      <a href={href} onClick={this.onClick}>
        {children}
      </a>
    );
  }
}

export default withRouter(LinkNoPrefetch);
```

**Reply by jiv-e** [reply]
This helped me to resolve my issue. Thanks!

**Comment by jiv-e** [comment]
PROBLEM: Unreliable page view logs
Not being able to disable the prefetching is especially problematic when I need to create a log event for every page view. This can be a requirement on some internal pages.

The idea is that admins could read from the logs which pages are viewed and by whom. If the user only hovers over the link it now creates a request. This can give a false information about what pages the user really visited.

**Reply by arch-user-france1** [reply]
just run a second request on the client that is only run when the page is really loaded

**Comment by kc980602** [comment]
After trying a few workaround, I decided to patch nextjs router directly

```
diff --git a/node_modules/next/dist/shared/lib/router/router.js b/node_modules/next/dist/shared/lib/router/router.js
index df2ec1d..e64e85f 100644
--- a/node_modules/next/dist/shared/lib/router/router.js
+++ b/node_modules/next/dist/shared/lib/router/router.js
@@ -1082,7 +1082,7 @@ class Router {
                     unstable_skipClientCache
                 };
                 const data = yield withMiddlewareEffects({
-                    fetchData: ()=>fetchNextData(fetchNextDataParams),
+                    fetchData: null,
                     asPath: resolvedAs,
                     locale: locale,
                     router: _this


```

**Comment by 2-fly-4-ai** [comment]
The prefetching should come with warnings and the next/link component should not be forced on you. The only reason I used it was that during build time, It would not allow me to proceed until I changed all internal <a> tags to a link component. This absolutely destroyed my serverless functions limit in 8 days and forced me to take down my projects due to the crazy costs that would have incurred. Even after disabling prefetch it still fetches every single link the person's mouse scrolls over while navigating down the page. So much extra tax on the backend slows the site down and costs you a bunch of extra money for nothing. 

*Edited to be a bit less obscene. I may have jumped the gun on this one. I still think they should warn newcomers and put a warning for sites with lots of links on the <Link> documentation. But they obviously know what they are doing when it comes to optimizing. It just leaves a pretty sour taste in the mouths of newcomers when they get stung with nasty bills. Prefetch behavior + WP jam stack is a total fail on ecom sites with lots of links. Especially since the behavior is different in dev vs production. It's a costly lesson to learn.

**Reply by elie222** [reply]
Yes. It's disappointing. Good chance they refund you for this first month btw. But the system is broken as you mention.

You switch off the auto prefetch. 

It's also that there's such a steep cliff between free and paid. First 1000gb is free then 55 dollars for every 100gb you go over. If you hit the limit you'll very quickly rack up a large bill.

**Reply by leerob** [reply]
> How many people have been ripped off an overcharged for this feature?

Note: It appears this feedback is related to Vercel, not Next.js.

> The prefetching should come with warnings and the next/link component should not be forced on you.

The default for `next/link` helps improve page transition performance, leading to faster applications – we still believe this is the correct default to provide a "single-page app experience". If you do not want to use prefetching, you can disable it on a per-link basis or globally by wrapping links for a component (there's a snippet above).

> The only reason I used it was that during build time, It would not allow me to proceed until I changed all internal tags to a link component.

This sounds like your ESLint rules, you can change this if you would like.

> So much extra tax on the backend slows the site down and costs you a bunch of extra money for nothing.

You can use the `a` tag if you prefer – you don't have to use `next/link` if you are worried about your backend performance. Although, I would say this is likely going to be an issue regardless. When the user clicks on the link, now users are staring at a browser loading spinner for 4 seconds. I would recommend trying to address the slow backend.

> I for one am very disappointed in Vercel and their endless crusade to monetize the dumb things.

I'm sorry you had a poor experience here. As mentioned above, we still believe the design of `next/link` to be correct. It sounds like in this instance should would prefer using `a` instead.

> I'm sure there are many large companies and individuals that have run tests and come to the same conclusion.

The majority of customers I talk to are very happy with `next/link` and prefetching.

> First 1000gb is free then 55 dollars

This is no longer the case – it's now $40 – and we're continually re-evaluating pricing based on feedback, so thank you!

**Reply by 2-fly-4-ai** [reply]
I think you are wrong, mate. Why should I optimize my backend to be smashed by a million requests as people scroll down the page? WordPress backends are not ideal, but they're the gateway drug to full-stack software engineering for many newcomers.

Pretty hilarious that they refused to refund this. It speaks volumes about their business practices. We've all seen those absolutely ridiculous charges racked up for the weirdest things.

It's not intuitive at all. And I can assure you that many devs have been hustled by this.

Thankfully, I now understand it as default behavior and can work around it. Very costly lesson, though, as it absolutely murdered my rankings at the time.

This is one of those "overoptimized" and unintuitive features that catch people off guard. In a normal app, it totally makes sense to prefetch like this.  But people are encouraged to build full sites with this, which is what I did at the time.

It took quite a while to even figure out what was going on. Had I not checked my bill 3 days into the month??? I don't even wanna imagine what would have happened. They could have straight-up bankrupted me at the time.

It's funny to laugh about in retrospect, but damn. I would love to see the numbers on how many people got stung by this and how much money was accumulated from just this one issue.

I haven't run into a single issue close to as expensive as this for the last year+ of development with next. So that says to me, this is actually a huge blunder in terms of being at least intuitive to newer developers.

It also put me off using Vercel for quite some time. Pretty bad start to a business relationship.

If you go back now and read your response, it doesn't really make sense. Of course, most customers would be happy with it as they are not running into issues with the types of apps they built.

----------------------

**Reply by azizoid** [reply]
I wander how stupid was a person who brought the prefetching idea in the table. And how did others agreed on that. Today internet speed is fast enough to load any page after the click. The the hell needs prefetching? Are we in 90s or what?

**Reply by 2-fly-4-ai** [reply]
Nextjs is trying to be as perfomant as possible out of the box. But this one issue really grinds my gears. And I absolutely loathe the thought of a billion-dollar+ company pinching pennies from new devs.

The next team has some of the smartest people in the world working there, and there is a reason for the things they do. But I don't think it's worth tainting the relationship with new customers over something so trivial.

It also sucks that you have to signup for the 20$ plan to test any website that has any sort of outward affiliate link. I could have definitely avoided this issue if that hadn't been the case. It's like the perfect recipe for disaster(or profit).

But alas. Goodluck to new devs running into this issue. It sucks. They should have an auto refund policy on this.

**Reply by unleashit** [reply]
One reason (exercise of the reader to speculate on the possible $ incentive) I think aggressive prefetching came about with Next.js because of how aggressively they code split. If the balance were even a bit more towards slightly larger initial bundle sizes, client navigation could be smoother and quicker with less need to prefetch (though I think it's useful selectively). I've come to realize that this is actually the heart of many problems with the way Next was designed. But considering the hivemind is moving away from SPAs and to treat the client as dumb as possible (never mind that mobile phones these days are super computers of 20 yrs ago), it seems like we'll have to get used to it unless you want to roll your own.

**Reply by daveycodez** [reply]
> > How many people have been ripped off an overcharged for this feature?
> 
> Note: It appears this feedback is related to Vercel, not Next.js.
> 
> > The prefetching should come with warnings and the next/link component should not be forced on you.
> 
> The default for `next/link` helps improve page transition performance, leading to faster applications – we still believe this is the correct default to provide a "single-page app experience". If you do not want to use prefetching, you can disable it on a per-link basis or globally by wrapping links for a component (there's a snippet above).
> 
> > The only reason I used it was that during build time, It would not allow me to proceed until I changed all internal tags to a link component.
> 
> This sounds like your ESLint rules, you can change this if you would like.
> 
> > So much extra tax on the backend slows the site down and costs you a bunch of extra money for nothing.
> 
> You can use the `a` tag if you prefer – you don't have to use `next/link` if you are worried about your backend performance. Although, I would say this is likely going to be an issue regardless. When the user clicks on the link, now users are staring at a browser loading spinner for 4 seconds. I would recommend trying to address the slow backend.
> 
> > I for one am very disappointed in Vercel and their endless crusade to monetize the dumb things.
> 
> I'm sorry you had a poor experience here. As mentioned above, we still believe the design of `next/link` to be correct. It sounds like in this instance should would prefer using `a` instead.
> 
> > I'm sure there are many large companies and individuals that have run tests and come to the same conclusion.
> 
> The majority of customers I talk to are very happy with `next/link` and prefetching.
> 
> > First 1000gb is free then 55 dollars
> 
> This is no longer the case – it's now $40 – and we're continually re-evaluating pricing based on feedback, so thank you!

How come the Next.js documentation website doesn't use prefetching then?

**Comment by unleashit** [comment]
There's another option that the framework could provide that could provide a balance to this issue. Rather than prefetch all the things, why not provide the developer with a little control over the way things are bundled. In the pre-framework isomorphic days, it used to be simple to control code splitting. Use dynamic import when you want it, and don't when you don't. Per page resources is safe and a good starting point, but it is a valid performance trade off decision to consider an initial bundle with a few common routes. In my experience, that's a better overall UX for the client and doesn't require hitting the server needlessly.

Of course that goes against Google's holy grail of of small initial payload over everything else no matter what. So you'd have to go easy with this power to please the SEO gods (and yeah of course mobile), but in my experience adding 25k or so to gain a handful of "pre" prefetched routes doesn't hurt "nearly" as bad as adding 2mb+ of ad and tracking libraries.

**Comment by khuezy** [comment]
So the new Link prefetch behavior seems to be broken? If you set `prefetch={false}`, then it doesn't even cache the request. So if you click on a static link multiple times, it keeps sending a request to the server. Why doesn't it cache the first request and reuse subsequent ones?

**Reply by daveycodez** [reply]
$$$

**Comment by arch-user-france1** [comment]
Just write your own link tag, it's easy:

```
// components/main.js
import { useRouter } from "next/router"
export function Link({ href, prefetch, hoverPrefetch, Components }) {
  if (prefetch) {
    Router.prefetch(href)
  }
  if (hoverPrefetch) {
    // implement with useEffect
  }
  return (
  <a href={href}>{...Components}</a>
  )
}
```
Note that this is an example and probably has one or two mistakes.

**Comment by brogr** [comment]
From what I can tell, v13.2.4 (app directory) doesn't prefetch on hover.
I've solved it like this for links that should only prefetch on hover but with a little bit of a delay:

```
"use client";

import { useState } from "react";
import { useRouter } from "next/navigation";

/* 
 * Prefetch link with delay on hover instead of instantly or when in view (default next/link behaviour)
*/
const DELAY = 1500;
export default function LinkHoverPrefetch({
	children,
	href,
	className,
}) {
	const router = useRouter();
  const [delayed, setDelayed] = useState(false);
  const [prefetched, setPrefetched] = useState(false);

	const onMouseOver = (e) => {
    if (!prefetched) {
      const timeout = setTimeout(function () {
				if (!prefetched) {
					console.log("delayed hover prefetch", href);
					router.prefetch(href);
					setPrefetched(true);
				}
			}, DELAY);
			setDelayed(timeout);
		}
	};

	const onMouseOut = (e) => {
    if (delayed) {
      console.log("clear timeout, abort hover prefetch", href);
			clearTimeout(delayed);
			setDelayed(false);
		};
	};

	return (
		<a
			href={href}
			className={className}
			onMouseOver={onMouseOver}
			onMouseOut={onMouseOut}
		>
			{children}
		</a>
	);
}
```

**Reply by wimbarelds** [reply]
With `prefetch={false}`, I'm also not seeing prefetch on hover with Next 13.4.3.

Granted, the [API docs](https://nextjs.org/docs/pages/api-reference/components/link#prefetch) don't say that it should prefetch on hover, however the [Code-docs do say it will still prefetch on hover](https://github.com/vercel/next.js/blob/canary/packages/next/src/client/link.tsx#L74).

**Reply by azizoid** [reply]
the question is why. why the f they add a feature without letting us to disable it

**Reply by youminkim** [reply]
I found that this is still same case in 14.0.3. with `prefetch=false`, no prefetch happening on hover. The code doc is outdated

**Comment by unleashit** [comment]
I can't speak for why they didn't in the past, but as someone mentioned above it's now possible. What I'd do is make a little wrapper over Link and change the default prefetch to false. Something like:

```typescript jsx
import NextLink, { type LinkProps } from 'next/link';

type CustLink = LinkProps & { children: React.ReactNode };

export function Link({ href, prefetch = false, children, ...rest }: CustLink) {
  return (
    <NextLink href={href} prefetch={prefetch} {...rest}>
      {children}
    </NextLink>
  );
}
```

**Comment by JohnAlbin** [comment]
@unleashit Thanks for the code! I was getting a Type error when adding a `className` to that component and fixed it with the code below.

`next/link` supports Ref, so if y'all want to add `Ref` forwarding, the component would look like this:

```javascript
// components/Link.tsx

import { forwardRef, type AnchorHTMLAttributes, type ReactNode } from "react"
import NextLink, { type LinkProps as NextLinkProps } from "next/link"

type LinkProps = NextLinkProps &
  Omit<AnchorHTMLAttributes<HTMLAnchorElement>, keyof NextLinkProps> & {
    children?: ReactNode
  }

export const Link = forwardRef<HTMLAnchorElement, LinkProps>(
  function LinkWithRef(
    {
      // Turn next/link prefetching off by default.
      // @see https://github.com/vercel/next.js/discussions/24009
      prefetch = false,
      ...rest
    },
    ref
  ) {
    return <NextLink prefetch={prefetch} {...rest} ref={ref} />
  }
)
```

**Reply by dmytrochygarov** [reply]
Thank you for your code! I didn't need the forward ref, so here is updated simplified code:

```
import { type AnchorHTMLAttributes, type ReactNode } from "react";
import NextLink, { type LinkProps as NextLinkProps } from "next/link";

type LinkProps = NextLinkProps &
  Omit<AnchorHTMLAttributes<HTMLAnchorElement>, keyof NextLinkProps> & {
    children?: ReactNode;
  };

export const Link = ({ prefetch = false, ...rest }: LinkProps) => {
  return <NextLink prefetch={prefetch} {...rest} />;
};
```

**Comment by g-wozniak** [comment]
Took me two hours to realise why my users are automatically logged off upon logging in to the admin panel; because one of the links was "Sign off" without damn prefetch. Now, when they hover over on "sign off" without the intention of being logged off, it will still happen unless I use <a href>.

**Reply by Basemm** [reply]
Aside from prefetching issues, sign out should be a POST request not a GET request. Imagine users sharing that link and forcing everyone to signout!

**Comment by kc980602** [comment]
I'm using v14.1.0 pages router with `getInitialProps`, it's still prefetch on hover. 

The easiest way to disable prefetching is to patch nextjs source code, forget about the workarounds mentioned above, just patch it!

**patches/next+14.1.0.patch**

```
diff --git a/node_modules/next/dist/shared/lib/router/router.js b/node_modules/next/dist/shared/lib/router/router.js
index 904f2c5..c949658 100644
--- a/node_modules/next/dist/shared/lib/router/router.js
+++ b/node_modules/next/dist/shared/lib/router/router.js
@@ -1110,7 +1110,7 @@ class Router {
                 isBackground
             };
             let data = isQueryUpdating && !isMiddlewareRewrite ? null : await withMiddlewareEffects({
-                fetchData: ()=>fetchNextData(fetchNextDataParams),
+                fetchData: null,
                 asPath: isNotFound ? "/404" : resolvedAs,
                 locale: locale,
                 router: this
@@ -1395,23 +1395,7 @@ class Router {
             }
         }
         const data = process.env.__NEXT_MIDDLEWARE_PREFETCH === "strict" ? null : await withMiddlewareEffects({
-            fetchData: ()=>fetchNextData({
-                    dataHref: this.pageLoader.getDataHref({
-                        href: (0, _formaturl.formatWithValidation)({
-                            pathname: originalPathname,
-                            query
-                        }),
-                        skipInterpolation: true,
-                        asPath: resolvedAs,
-                        locale
-                    }),
-                    hasMiddleware: true,
-                    isServerRender: this.isSsr,
-                    parseJSON: true,
-                    inflightCache: this.sdc,
-                    persistCache: !this.isPreview,
-                    isPrefetch: true
-                }),
+            fetchData: null,
             asPath: asPath,
             locale: locale,
             router: this
@@ -1443,19 +1427,7 @@ class Router {
         }
         await Promise.all([
             this.pageLoader._isSsg(route).then((isSsg)=>{
-                return isSsg ? fetchNextData({
-                    dataHref: (data == null ? void 0 : data.json) ? data == null ? void 0 : data.dataHref : this.pageLoader.getDataHref({
-                        href: url,
-                        asPath: resolvedAs,
-                        locale: locale
-                    }),
-                    isServerRender: false,
-                    parseJSON: true,
-                    inflightCache: this.sdc,
-                    persistCache: !this.isPreview,
-                    isPrefetch: true,
-                    unstable_skipClientCache: options.unstable_skipClientCache || options.priority && !!process.env.__NEXT_OPTIMISTIC_CLIENT_CACHE
-                }).then(()=>false).catch(()=>false) : false;
+                return false;
             }),
             this.pageLoader[options.priority ? "loadPage" : "prefetch"](route)
         ]);
```

**Comment by rusakovic** [comment]
This s***t is still not solved! 
My website perfomance is so bad, and the core is PREFETCHING.

Will try 
```
You can opt out of prefetching using <Link prefetch={false}>.
```

**Comment by blackbing** [comment]
Here is how I fix it simply. 
Add this script to overwrite the Link Component in _app.tsx

```jsx
import Link from 'next/link';
import type { LinkProps } from 'next/link';

const OriginalLink = Link;

Object.defineProperty(Link, 'default', {
  configurable: true,
  writable: true,
  value: (props: LinkProps) => {
    return <OriginalLink {...props} prefetch={props.prefetch ?? false} />;
  },
});
```

**Reply by IanGraingerGMSL** [reply]
> Here is how I fix it simply. Add this script to overwrite the Link Component in _app.tsx
> 
> ```js
> import Link from 'next/link';
> import type { LinkProps } from 'next/link';
> 
> const OriginalLink = Link;
> 
> Object.defineProperty(Link, 'default', {
>   configurable: true,
>   writable: true,
>   value: (props: LinkProps) => {
>     return <OriginalLink {...props} prefetch={props.prefetch ?? false} />;
>   },
> });
> ```

I assume this file was ignored by my Next build because this didn't work for me ☹️ .

**Comment by BertrandBordage** [comment]
If you want to disable all prefetching (hover included), here is a drop-in replacement for `Link` with all its features except:
- `prefetch`
- `passHref`
- `legacyBehavior`

**Replace next/navigation with next/router** if you use the pages router.

```tsx
"use client";

import {
  forwardRef,
  useCallback,
  type AnchorHTMLAttributes,
  type ReactNode,
} from "react";
import { type LinkProps } from "next/link";
import { useRouter } from "next/navigation";

export type NextLinkProps = AnchorHTMLAttributes<HTMLAnchorElement> &
  Pick<
    LinkProps,
    "href" | "as" | "locale" | "replace" | "shallow" | "scroll" | "onNavigate"
  > & {
    children?: ReactNode;
  };

const NextLink = forwardRef<HTMLAnchorElement, NextLinkProps>(
  function LinkWithRef(
    {
      href,
      as,
      locale,
      replace,
      shallow,
      scroll,
      onNavigate,
      onClick,
      ...rest
    },
    ref,
  ) {
    const router = useRouter();

    const routeOnClick = useCallback<
      React.MouseEventHandler<HTMLAnchorElement>
    >(
      (event) => {
        if (onClick) {
          onClick(event);
        }

        if (event.defaultPrevented) {
          return;
        }

        if (onNavigate) {
          let isDefaultPrevented = false;

          onNavigate({
            preventDefault: () => {
              isDefaultPrevented = true;
            },
          });

          if (isDefaultPrevented) {
            return;
          }
        }

        event.preventDefault();
        if (replace) {
          router.replace(href, as, { shallow, locale, scroll });
        } else {
          router.push(href, as, { shallow, locale, scroll });
        }
      },
      [router, href, as, locale, replace, shallow, scroll, onNavigate, onClick],
    );

    return <a ref={ref} href={href} onClick={routeOnClick} {...rest} />;
  },
);

export default NextLink;
```

**Comment by HasanMothaffar-OSG** [comment]
So this is enabled by default just so hosting providers like Vercel can money off of this feature? It should be disabled by default and only enabled by choice

**Reply by azizoid** [reply]
Unfortunately there is no free cheese

**Reply by unleashit** [reply]
Sure there's free cheese. You don't even have to roll your own, there's plenty of great FOSS solutions that work fine. Some of them are older and less complicated. Oh no. The situation is that Vercel used hype and social media influencing to trick the industry into Next becoming the bait that doesn't stop giving. Way before even Next 13, around the time they introduced getServersideProps, it was obvious they were going for maximum unnecessary server requests.  The best solution to this thread and others is to use something else.

**Comment by BarishNamazov** [comment]
I almost was got by this cash trap too. For me, Next.js did two, not one, requests per page too -- one for the SEO metadata. No idea why it couldn't combine two <2KB requests into one. Fully disabling prefetching is the way here.

**Comment by kachkaev** [comment]
The problem with a global config is that it introduces a _mode_ in DX. Modes don't play well with human cognition. If I contribute to multiple projects, I have to keep in mind what default `prefetch` value each of them has. The brain won't handle this well which will lead to mistakes and distraction.

I have a project where I don't want to do prefetching in the vast majority of cases. Creating a custom 'wrapper' Link component could do the job but would come with a footgun. It would not save devs or AI agents from accidentally using `next/link` directly, thus ignoring the desired _mode_.

My solution is based on leveraging [`no-restricted-syntax`](https://eslint.org/docs/latest/rules/no-restricted-syntax) rule in ESLint:

```ts
// eslint.config.ts

const myEslintConfig = defineConfig(
  { "...": "..." },

  {
    rules: {
      "...": "...",

      "no-restricted-syntax": [
        "warn",

        { "...": "..." },

        {
          selector:
            "JSXAttribute[name.name=prefetch][value.expression.value=true]",
          message:
            "We generally avoid prefetching in next/link instances to avoid excessive API calls in proxy.ts. If you want to make an exception, add an eslint-ignore comment with an explanation.",
        },
        {
          selector:
            "JSXOpeningElement[name.name=Link]:not(:has(JSXAttribute[name.name=prefetch]))",
          message:
            "Please add prefetch={false} to this next/link instance or use <a> for an external link. If you want to make an exception, please add an eslint-ignore comment with an explanation.",
        },
      ],
    },
  },
);
```

```tsx
// example.tsx

import Link from "next/link";

function Example() {
  return (
    <>
      {/* ⚠️ ESLint warns: "Please add prefetch={false} to this next/link instance ..." */}
      <Link href="/">Home</Link>

      {/* ⚠️ ESLint warns: "We generally avoid prefetching in next/link instances ..." */}
      <Link href="/" prefetch={true}>
        Home
      </Link>

      {/* ✅ ESLint passes - not prefetching */}
      <Link href="/" prefetch={false}>
        Home
      </Link>

      {/* ✅ ESLint passes - prefetching, but explaining why */}
      {/* eslint-disable-next-line no-restricted-syntax -- we're making an exception because ... */}
      <Link href="/example" prefetch={true}>
        Example
      </Link>
    </>
  );
}
```

With this approach, both humans and AI agents stay in the [pit of success](https://blog.codinghorror.com/falling-into-the-pit-of-success/). There is no need to remember or learn anything new when adding new links. 

Hope it helps! 🙌

**Comment by fabriziofeitosa** [comment]
I thought @kachkaev idea was excellent, and since I use [Biome](https://biomejs.dev/), I decided to adapt it. Here's what I did:

1. I created a plugin `scripts/biome/requireNextLinkPrefetchFalse.grit`

```
engine biome(1.0)
language js(typescript, jsx)

or {
  `<Link $props />` as $link where {
    not $props <: contains `prefetch={false}`,
    register_diagnostic(span=$link, message="Every <Link> must declare prefetch={false}.")
  },
  `<Link $props>$...</Link>` as $link where {
    not $props <: contains `prefetch={false}`,
    register_diagnostic(span=$link, message="Every <Link> must declare prefetch={false}.")
  }
}
```

2. In `biome.json`:

```json
  "linter": {
    "enabled": true,
  },
  "plugins": ["./scripts/biome/requireNextLinkPrefetchFalse.grit"],
```


And if you want to ignore the rule, just use:

```tsx
// biome-ignore lint/plugin/requireNextLinkPrefetchFalse: intentional exception in this case
<Link href={forumUrl} target="_blank" rel="noopener noreferrer">
  Forum
</Link>
```

**Comment by landsman** [comment]
There is still no native configuration for this? 👀

**Comment by xdatabox** [comment]
Next.js currently does not provide a global configuration to completely disable prefetching.

However, you can control it per-link using the `prefetch` prop:

<Link href="/page" prefetch={false}>
  Go to page
</Link>

For larger apps, a common workaround is to create a custom Link wrapper component that disables prefetch by default.

Example:

import Link from "next/link";

export default function CustomLink(props) {
  return <Link {...props} prefetch={false} />;
}

This gives you centralized control without modifying every individual Link.

In development mode, prefetching behavior can also appear more aggressive, but in production it is optimized differently.

**Reply by fabriziofeitosa** [reply]
This solution has already been proposed in a reply in this thread.

**Comment by icyJoseph** [comment]
We are nearly there... https://nextjs.org/blog/next-16-3-instant-navigations#rethinking-prefetching 🎉 

If you do try out the 16.3 preview, please do provide feedback in this [discussion](https://github.com/vercel/next.js/discussions/95130).

**Reply by dimitriBouteille** [reply]
Hallelujah 🚀
